clients/python/girder_client/__init__.py

```diff
@@ -749,41 +793,62 @@ def _uploadContents(self, uploadObj, stream, size, progressCallback=None):
         """
         offset = 0
         uploadId = uploadObj['_id']
-        while True:
-            chunk = stream.read(min(self.MAX_CHUNK_SIZE, (size - offset)))
 
-            if not chunk:
-                break
+        with self.progressReporterCls(label=uploadObj['name'], length=size) as reporter:
 
-            if isinstance(chunk, six.text_type):
-                chunk = chunk.encode('utf8')
+            # Used for fast non-multipart upload
+            class _ProgressBytesIO(six.BytesIO):
+                def read(self, _size):
+                    _chunk = six.BytesIO.read(self, _size)
+                    reporter.update(len(_chunk))
+                    return _chunk
 
-            if self.getServerVersion() >= ['2', '2']:
-                uploadObj = self.post(
-                    'file/chunk?offset=%d&uploadId=%s' % (offset, uploadId), data=chunk)
-            else:
-                # Prior to version 2.2 the server only supported multipart uploads
-                parameters = {
-                    'offset': offset,
-                    'uploadId': uploadId
-                }
+            # Used for deprecated multipart upload
+            class _ProgressMultiPartEncoder(MultipartEncoder):
+                def read(self, _size):
+                    _chunk = MultipartEncoder.read(self, _size)
+                    reporter.update(len(_chunk))
+                    return _chunk
```

Comment:
Could we move these classes out of the function and up to the module level instead? Makes them easy to mock, and slightly less overhead when running this function.